File: _java-archive/acquisition-service/src/main/java/dev/kritchalach/acquisition/domain/Acquisition.java
package dev.kritchalach.acquisition.domain;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

/**
 * Aggregate Root: Acquisition
 * ครอบคลุม lifecycle การเข้าซื้อ property ของบริษัท ตั้งแต่รับ survey, ขอ CEO อนุมัติ,
 * จนถึงร่าง willing contract ทุก Command ต้องผ่าน Acquisition Root
 */
@Entity
@Table(name = "acquisitions")
public class Acquisition {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID acquisitionId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private AcquisitionStatus status;

    @Version
    private int version;

    @Column(nullable = false)
    private Instant createdAt;

    private Instant approvedAt;

    // Ref only — เก็บแค่ ID เพื่อ trace กลับไปยัง upstream domain
    @Column(nullable = false)
    private UUID surveyId;

    @Column(nullable = false)
    private UUID propertyId;

    @Embedded
    private PropertySurvey survey;

    @Embedded
    private SellerInfo seller;

    // Entity ภายใน Aggregate — สร้างได้หลังจาก CEO approve
    @OneToOne(cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @JoinColumn(name = "willing_contract_id")
    private WillingContract willingContract;

    protected Acquisition() {}

    /**
     * Factory: เริ่ม acquisition จาก property.surveyed event
     * Command: ReceiveSurvey
     */
    public static Acquisition fromSurvey(UUID surveyId, UUID propertyId,
                                          PropertySurvey survey, SellerInfo seller) {
        Acquisition acq = new Acquisition();
        acq.surveyId = surveyId;
        acq.propertyId = propertyId;
        acq.survey = survey;
        acq.seller = seller;
        acq.status = AcquisitionStatus.SURVEYED;
        acq.createdAt = Instant.now();
        return acq;
    }

    /**
     * Command: RequestApproval
     * Legal ขอ CEO อนุมัติให้เข้าซื้อ property
     */
    public void requestApproval() {
        if (status != AcquisitionStatus.SURVEYED) {
            throw new IllegalStateException(
                    "Cannot request approval — status must be SURVEYED but was " + status);
        }
        this.status = AcquisitionStatus.APPROVAL_REQUESTED;
    }

    /**
     * Command: ApproveAcquisition
     * รับการอนุมัติจาก CEO
     */
    public void approve() {
        if (status != AcquisitionStatus.APPROVAL_REQUESTED) {
            throw new IllegalStateException(
                    "Cannot approve — status must be APPROVAL_REQUESTED but was " + status);
        }
        this.status = AcquisitionStatus.APPROVED;
        this.approvedAt = Instant.now();
    }

    /**
     * Command: DraftWillingContract
     * Factory method สำหรับสร้าง WillingContract ภายใน aggregate
     * Business Rule: ต้องอนุมัติแล้วเท่านั้นจึงร่างสัญญาจะซื้อจะขายได้
     */
    public WillingContract draftWillingContract(UUID templateId, String fileUrl, BigDecimal agreedPrice) {
        if (status != AcquisitionStatus.APPROVED) {
            throw new IllegalStateException(
                    "Cannot draft willing contract — status must be APPROVED but was " + status);
        }
        this.willingContract = new WillingContract(templateId, fileUrl, agreedPrice);
        this.status = AcquisitionStatus.CONTRACT_DRAFTED;
        return this.willingContract;
    }

    public UUID getAcquisitionId() { return acquisitionId; }
    public AcquisitionStatus getStatus() { return status; }
    public int getVersion() { return version; }
    public Instant getCreatedAt() { return createdAt; }
    public Instant getApprovedAt() { return approvedAt; }
    public UUID getSurveyId() { return surveyId; }
    public UUID getPropertyId() { return propertyId; }
    public PropertySurvey getSurvey() { return survey; }
    public SellerInfo getSeller() { return seller; }
    public WillingContract getWillingContract() { return willingContract; }
}

File: acquisition-service/src/domain/Acquisition.ts
import { v4 as uuidv4 } from 'uuid';
import { AcquisitionStatus } from './AcquisitionStatus.js';
import { PropertySurvey } from './PropertySurvey.js';
import { SellerInfo } from './SellerInfo.js';
import { WillingContract } from './WillingContract.js';

/**
 * Aggregate Root: Acquisition
 * ครอบคลุม lifecycle การเข้าซื้อ property:
 *   SURVEYED → APPROVAL_REQUESTED → APPROVED → CONTRACT_DRAFTED
 * ทุก Command ต้องผ่าน Acquisition Root
 */
export class Acquisition {
  acquisitionId!: string;
  status!: AcquisitionStatus;
  version: number = 0;
  createdAt!: string;
  approvedAt?: string;
  surveyId!: string;
  propertyId!: string;
  survey!: PropertySurvey;
  seller!: SellerInfo;
  willingContract?: WillingContract;
  // Rich property + unit metadata captured from CEO's survey
  propertyDeveloper?: string;
  propertyName?: string;
  propertyType?: string;
  propertyCode?: string;
  city?: string;
  currency?: string;
  registration?: string;
  createdBy?: string;
  unitId?: string;
  unitCode?: string;
  unitArea?: number;
  bedroomType?: string;
  unitAddress?: string;
  bathrooms?: number;
  view?: string;
  furniture?: string;
  facility?: string;
  pictureUrls?: string[];
  cost?: number;
  minSalePrice?: number;
  price?: number;
  saleTeamLead?: string;
  commission?: number;
  externalStatus?: string;

  /**
   * Factory: เริ่ม acquisition จาก property.surveyed event
   */
  static fromSurvey(input: {
    surveyId: string;
    propertyId: string;
    survey: PropertySurvey;
    seller: SellerInfo;
  }): Acquisition {
    const a = new Acquisition();
    a.acquisitionId = uuidv4();
    a.surveyId = input.surveyId;
    a.propertyId = input.propertyId;
    a.survey = input.survey;
    a.seller = input.seller;
    a.status = AcquisitionStatus.SURVEYED;
    a.createdAt = new Date().toISOString();
    return a;
  }

  /**
   * Command: RequestApproval
   */
  requestApproval(): void {
    if (this.status !== AcquisitionStatus.SURVEYED) {
      throw new Error(`Cannot request approval — status must be SURVEYED but was ${this.status}`);
    }
    this.status = AcquisitionStatus.APPROVAL_REQUESTED;
  }

  /**
   * Command: ApproveAcquisition
   */
  approve(): void {
    if (this.status !== AcquisitionStatus.APPROVAL_REQUESTED) {
      throw new Error(`Cannot approve — status must be APPROVAL_REQUESTED but was ${this.status}`);
    }
    this.status = AcquisitionStatus.APPROVED;
    this.approvedAt = new Date().toISOString();
  }

  /**
   * Command: DraftWillingContract
   * Business Rule: ต้องอนุมัติแล้วเท่านั้นจึงร่างสัญญาจะซื้อจะขายได้
   */
  draftWillingContract(templateId: string, fileUrl: string, agreedPrice: number): WillingContract {
    if (this.status !== AcquisitionStatus.APPROVED) {
      throw new Error(`Cannot draft willing contract — status must be APPROVED but was ${this.status}`);
    }
    this.willingContract = WillingContract.create(templateId, fileUrl, agreedPrice);
    this.status = AcquisitionStatus.CONTRACT_DRAFTED;
    return this.willingContract;
  }
}
